checkpoint/orbax/checkpoint/_src/serialization/replica_slices.py

```diff
@@ -137,30 +138,28 @@ def to_fragments(self) -> fragments.Fragments:
     return result
 
 
-def _hashable_slices(slices: Index) -> tuple[HashableSlice, ...]:
-  return tuple([(s.start, s.stop, s.step) for s in slices])
-
-
-def _create_replica_counts_builder(
-    arr: jax.Array,
-) -> Callable[[jax.Shard], int]:
-  """Produces a mapping from addressable shards to global replication count."""
-
-  counts = collections.defaultdict(int)
-  for index in arr.sharding.devices_indices_map(arr.shape).values():
-    counts[_hashable_slices(index)] += 1
-
-  return lambda shard: counts[_hashable_slices(shard.index)]
+def _sharding_num_replicas(
```

Comment:
Looks fine to me in theory, though I'm not that familiar with these APIs. Is there any way of relying on fewer private JAX APIs though? I'm a little cautious about introducing a bunch more (we do already have some, and that is a source of some concern).

Reply:
I don't think there's any reasonable way around interacting with XLA's (somewhat intimidating) HloShardings. But we could more or less duplicate part of `jax._src.op_shardings` ([these parts](https://github.com/jax-ml/jax/blob/bf0150bb22b2ed7986adf3762cb1bc555ed3fee8/jax/_src/op_shardings.py#L27-L53)) and avoid that particular JAX-internal dependency.